tailnet/coordinator.go

```diff
@@ -146,14 +168,36 @@ type core struct {
 	// nodes maps agent and connection IDs their respective node.
 	nodes map[uuid.UUID]*Node
 	// agentSockets maps agent IDs to their open websocket.
-	agentSockets map[uuid.UUID]*TrackedConn
+	agentSockets map[uuid.UUID]Enqueueable
 	// agentToConnectionSockets maps agent IDs to connection IDs of conns that
 	// are subscribed to updates for that agent.
-	agentToConnectionSockets map[uuid.UUID]map[uuid.UUID]*TrackedConn
+	agentToConnectionSockets map[uuid.UUID]map[uuid.UUID]Enqueueable
+
+	// clients holds a map of all clients connected to the coordinator. This is
+	// necessary because a client may not be subscribed into any agents.
+	clients map[uuid.UUID]Enqueueable
+	// clientsToAgents is an index of clients to all of their subscribed agents.
+	clientsToAgents map[uuid.UUID]map[uuid.UUID]struct{}
 
 	// agentNameCache holds a cache of agent names. If one of them disappears,
 	// it's helpful to have a name cached for debugging.
 	agentNameCache *lru.Cache[uuid.UUID, string]
+
+	// legacyAgents holda a mapping of all agents detected as legacy, meaning
+	// they only listen on codersdk.WorkspaceAgentIP. They aren't compatible
+	// with the new ServerTailnet, so they must be connected through
+	// wsconncache.
+	legacyAgents map[uuid.UUID]struct{}
+}
+
+type Enqueueable interface {
+	UniqueID() uuid.UUID
+	Enqueue(n []*Node) error
+	Name() string
+	Stats() (start, lastWrite int64)
+	Overwrites() int64
+	CoordinatorClose() error
+	Close() error
```

Comment:
I don't understand the distinction between `CoordinatorClose()` and `Close()` --- how are they different and why do we need both?

Reply:
There's sort-of a chicken and egg problem. When you want to close the individual queue, you expect it to be removed from the coordinator. When the coordinator is closing, you want the coordinator to close all of the queues, but you don't want the queues to reach back into the coordinator to close themselves and reenter the mutex.